keps/sig-apps/20190326-stateful-application-api.md

```diff
@@ -0,0 +1,790 @@
+
+	Spec   ApplicationSnapshotSpec   `json:"spec,omitempty"`
+	Status ApplicationSnapshotStatus `json:"status,omitempty"`
+}
+```
+
+`ApplicationSnapshotClass` is a cluster-scoped CRD that defines the snapshotter and common parameters for a class of `StatefulApplication`s. An `ApplicationSnapshot` object must specify the name of an `ApplicationSnapshotClass` object, which identifies the snapshotter and provide needed parameters to the snapshotter. The `ApplicationSnapshotClass` object also defines hooks for quiescing and unquiescing the application. Below is the definition of `ApplicationSnapshotClass`.
+
+```go
+type ApplicationSnapshotClass struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Snapshotter is the name of the snapshot controller that should be invoked.
+	Snapshotter string `json:"snapshotter"`
+
+	// VolumeSnapshotClassNames is mapping from StorageClass names to the names of the corresponding
+	// VolumeSnapshotClasses that should be used when creating VolumeSnapshots. The assumption is that
+	// snapshot of persistent volumes of the same StorageClass is handled by the same snapshotter
+	// designated by a VolumeSnapshotClass. The default VolumeSnapshotClass is used for PVCs with no
+	// StorageClass specified and for PVCs with a StorageClass not found in the mapping.
+	// +optional
+	VolumeSnapshotClassNames map[string]string `json:"volumeSnapshotClassNames,omitempty"`
+
+	// Parameters are opaque parameters to the snapsohtter.
+	// +optional
+	Parameters map[string]string `json:"parameters,omitempty"`
+
+  // PreHooks are an optional list of hooks to run to do application quiescing/flushing before
+	// taking an application snapshot. The hooks may return application-specific tags in JSON
+	// formatted key-value pairs, e.g., {“tag1”: “value1”, “tag2”: “value2”}. Each hook in the
+	// list is for a specific container of the application, and one container can have no more
+	// than one pre-hook.
+	// +optional
+	PreHooks []HookSpec `json:"preHooks,omitempty"`
+
+	// PostHooks are an optional list of hooks  to run to do application unquiescing after taking
+	// an application snapshot. Each hook in the list is for a specific container of the application,
+	// and one container of the application can have no more than one post-hook.
+	// +optional
+	PostHooks []HookSpec `json:"postHooks,omitempty"`
+
+	// DeletionPolicy is the policy on what happens to an ApplicationSnapshotContent resource and the
+	// local application snapshot it represents when the ApplicationSnapshot resource it is bound to
+	// is deleted.
+	// DeletionPolicy set in individual ApplicationSnapshotContent resources takes precedence over this.
+	// +optional
+	// +kubebuilder:validation:Enum=Delete,Retain
+	DeletionPolicy *DeletionPolicy `json:"deletionPolicy,omitempty"`
+}
+
+// HookSpec defines a hook for quiescing or unquiescing an application.
+type HookSpec struct {
```

Comment:
Why creating another HookSpec struct?  Can you use the ExecutionHook API object designed in this KEP (https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/20190120-execution-hook-design.md)?

Reply:
Discussed about this in today's review meeting. Can you please add some description to explain that ExecutionHook will be created based on HookSpec?

Reply:
I updated the KEP to use `ExecutionHookSpec` instead of the `HookSpec` struct.

Reply:
Actually stepping back a bit, I tried to run through the user flow if we use `ExecutionHookSpec` instead, and it seemed to need more typing from the users. I think it makes sense to keep the `HookSpec` struct here which is tailed for the needs here and is less verbose. I added descriptions to make it clear that an `ExecutionHook` object will be created at runtime for each hook based on information provided in the `HookSpec`.